trio/_highlevel_open_unix_stream.py

```diff
@@ -0,0 +1,42 @@
+import trio
+from trio._highlevel_open_tcp_stream import close_on_error
+from trio.socket import socket, SOCK_STREAM
+
+try:
+    from trio.socket import AF_UNIX
+    has_unix = True
+except ImportError:
+    has_unix = False
+
+__all__ = ["open_unix_socket"]
+
+
+async def open_unix_socket(filename,):
+    """Opens a connection to the specified
+    `Unix domain socket <https://en.wikipedia.org/wiki/Unix_domain_socket>`_.
+
+    You must have read/write permission on the specified file to connect.
+
+    Args:
+      filename (str or bytes): The filename to open the connection to.
+
+    Returns:
+      SocketStream: a :class:`~trio.abc.Stream` connected to the given file.
+
+    Raises:
+      OSError: If the socket file could not be connected to.
+      RuntimeError: If AF_UNIX sockets are not supported.
+    """
+    if not has_unix:
+        raise RuntimeError("Unix sockets are not supported on this platform")
```

Comment:
Huh, how to handle this is an interesting question.

One tradition says, unsupported operations should just be missing, so you can do `hasattr(trio, "open_unix_stream")` to check for whether it's there. The other approach is to do this. On the one hand, the missing approach is probably more traditional in Python, and it's what we do for OS-specific operations in `trio.hazmat`. OTOH, it might make it harder to get `trio.__all__` set up correctly and in a way that doesn't confuse static analysis tools, and we might need to switch how we handle OS-specific operations anyway to support pluggable backends. Hmm.

I'm going to hit post on this review so you can see the other comments, while I think about this a little more :-)

Reply:
I haven't come to any real conclusions here, but given the problems that autocompleters have when some functions are only conditionally defined (#314), and that this way callers get a better error message than `AttributeError: open_unix_stream`, I'm going to say let's leave this as is for now for now and we can revisit it later if we change our mind.